p2p/net/swarm/swarm_dial.go

```diff
@@ -364,98 +366,64 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	return swarmC, nil
 }
 
-func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs []ma.Multiaddr) (conn.Conn, error) {
-
-	// sort addresses so preferred addresses are dialed sooner
-	sort.Sort(AddrList(remoteAddrs))
-
-	// try to connect to one of the peer's known addresses.
-	// we dial concurrently to each of the addresses, which:
-	// * makes the process faster overall
-	// * attempts to get the fastest connection available.
-	// * mitigates the waste of trying bad addresses
+func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.Multiaddr) (conn.Conn, error) {
 	log.Debugf("%s swarm dialing %s %s", s.local, p, remoteAddrs)
 
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel() // cancel work when we exit func
 
-	conns := make(chan conn.Conn)
-	errs := make(chan error, len(remoteAddrs))
+	// use a single response type instead of errs and conns, reduces complexity *a ton*
+	respch := make(chan dialResult)
 
-	// dialSingleAddr is used in the rate-limited async thing below.
-	dialSingleAddr := func(addr ma.Multiaddr) {
-		// rebind chans in scope so we can nil them out easily
-		connsout := conns
-		errsout := errs
+	defaultDialFail := fmt.Errorf("failed to dial %s (default failure)", p)
+	exitErr := defaultDialFail
 
-		connC, err := s.dialAddr(ctx, p, addr)
-		if err != nil {
-			connsout = nil
-		} else if connC == nil {
-			// NOTE: this really should never happen
-			log.Errorf("failed to dial %s %s and got no error!", p, addr)
-			err = fmt.Errorf("failed to dial %s %s", p, addr)
-			connsout = nil
-		} else {
-			errsout = nil
-		}
-
-		// check parent still wants our results
+	var active int
+	for {
 		select {
-		case <-ctx.Done():
-			if connC != nil {
-				connC.Close()
+		case addr, ok := <-remoteAddrs:
+			if !ok {
+				remoteAddrs = nil
+				if active == 0 {
+					return nil, exitErr
+				}
+				continue
 			}
-		case errsout <- err:
-		case connsout <- connC:
-		}
-	}
 
-	// this whole thing is in a goroutine so we can use foundConn
-	// to end early.
-	go func() {
-		limiter := make(chan struct{}, 8)
-		for _, addr := range remoteAddrs {
-			// returns whatever ratelimiting is acceptable for workerAddr.
-			// may not rate limit at all.
-			rl := s.addrDialRateLimit(addr)
-			select {
-			case <-ctx.Done(): // our context was cancelled
-				return
-			case rl <- struct{}{}:
-				// take the token, move on
+			// limitedDial will start a dial to the given peer when
+			// it is able, respecting the various different types of rate
+			// limiting that occur without using extra goroutines per addr
+			s.limitedDial(ctx, p, addr, respch)
+			active++
+		case <-ctx.Done():
+			if exitErr == defaultDialFail {
```

Comment:
because its next to useless to the user. If we don't have any other better response, then we can use ctx.Err().